Name constraints + binding constraints

flomnes opened the pull request:
Add a name for some constraints (not all).

flomnes committed:
Binding constraints: export name into optimization problem

flomnes committed:
Fix for weekly binding constraints

flomnes committed:
Create struct PROBLEME_NOMME

flomnes committed:
Copy variable/constraint names

flomnes committed:
Clean construction of BC names, move before OPT_ChargerLaContrainte

flomnes committed:
Add comment in PROBLEME_ANTARES_A_RESOUDRE

flomnes committed:
Name constraint related to fictional load

flomnes committed:
Add hydro_level constraint name

flomnes committed:
Merge remote-tracking branch 'http/develop' into feature/var_constr_naming

flomnes committed:
Fix 2 code smells

flomnes committed:
Constify everything

flomnes commented on src/solver/utils/named_problem.h:
```diff
@@ -0,0 +1,13 @@
+                 const std::vector<std::string>& NomDesContraintes) : NomDesVariables(NomDesVariables), NomDesContraintes(NomDesContraintes) {}
+
+  const std::vector<std::string> NomDesVariables;
+  const std::vector<std::string> NomDesContraintes;
```
Add consistency check functions `this->NombreDeVariables == NomDesVariables.size()`

flomnes replied:
done

flomnes commented on src/solver/utils/ortools_utils.cpp:
```diff
@@ -129,13 +142,16 @@ MPSolver* convert_to_MPSolver(PROBLEME_A_RESOUDRE* problemeAResoudre)
                       problemeAResoudre->Xmax,
                       problemeAResoudre->CoutLineaire,
-                      problemeAResoudre->NombreDeVariables);
+                      problemeAResoudre->NombreDeVariables,
```
Do not provide an empty `std::vector`

flomnes reviewed (commented):
Add .md on how constraints are named

flomnes commented on src/solver/utils/named_problem.h:
```diff
@@ -0,0 +1,13 @@
+
+struct PROBLEME_NOMME : public PROBLEME_SIMPLEXE {
+  PROBLEME_NOMME(const std::vector<std::string>& NomDesVariables,
+                 const std::vector<std::string>& NomDesContraintes) : NomDesVariables(NomDesVariables), NomDesContraintes(NomDesContraintes) {}
```
`Copy of a std::vector<std::string>`. Is it necessary or can we just keep a reference ?

flomnes replied:
const ref

flomnes commented on src/solver/simulation/sim_structure_probleme_economique.h:
```diff
@@ -166,6 +167,7 @@ typedef struct
     int* PaysDuPalierDispatch;
     int* NumeroDuPalierDispatch;
     int* OffsetTemporelSurLePalierDispatch;
+    Antares::Data::ConstraintName NomDeLaContrainteCouplante;
```
`const char*` ? We add a dependency here

guilpier-code replied:
Indeed, why not define a string here (instead of a **ConstraintName**) ? (and remove the dependency)
No trouble to change this apparently

flomnes replied:
I'll use `const char*` to keep this structure in a "C-style"

flomnes reviewed (commented):
Performance ?

flomnes commented on src/solver/optimisation/opt_appel_solveur_lineaire.cpp:
```diff
@@ -98,9 +98,11 @@ bool OPT_AppelDuSimplexe(PROBLEME_HEBDO* ProblemeHebdo, uint numSpace, int NumIn
+
     PROBLEME_SPX* ProbSpx;
     ProblemeAResoudre = ProblemeHebdo->ProblemeAResoudre;
+    PROBLEME_NOMME Probleme(ProblemeAResoudre->NomDesVariables,
```
`PROBLEME_SIMPLEXE_NOMME`

flomnes replied:
Done

guilpier-code commented on src/solver/optimisation/opt_construction_matrice_des_contraintes_cas_lineaire.cpp:
```diff
@@ -285,8 +287,11 @@ void OPT_ConstruireLaMatriceDesContraintesDuProblemeLineaire(PROBLEME_HEBDO* Pro
             CorrespondanceCntNativesCntOptim->NumeroDeContraintePourEviterLesChargesFictives[Pays]
               = ProblemeAResoudre->NombreDeContraintes;
 
+            NomDeLaContrainte
```
ts = timeStepInYear, right ?
If so, please change its name to something more explicit

flomnes replied:
```cpp
int ts = study.runtime->weekInTheYear[numSpace] * 168 + Pdt;
```

Right, name must be changed.

guilpier-code commented on src/solver/utils/ortools_wrapper.h:
```diff
@@ -1,17 +1,15 @@
 #ifndef __ORTOOLS_WRAPPER__
```
This file is not in the IDE solution generated by cmake. It will not be searched in case of search over the entire solution.
Same thing for **ortools_utils.h**

flomnes replied:
ok

guilpier-code commented on src/libs/antares/study/runtime/runtime.h:
```diff
@@ -93,6 +93,7 @@ class BindingConstraintRTI
     int* clusterOffset;
     long* clusterIndex;
     long* clustersAreaIndex;
+    ConstraintName name;
```
why not a string ?

flomnes replied:
I'll be using a `const char*`

flomnes replied:
I'll be using a `std::string`. Using `const char*` would force us to manage memory (deep copies are needed).

guilpier-code commented on src/solver/optimisation/opt_alloc_probleme_a_optimiser.cpp:
```diff
@@ -258,6 +261,9 @@ void OPT_LiberationMemoireDuProblemeAOptimiser(PROBLEME_HEBDO* ProblemeHebdo)
         MemFree(ProblemeAResoudre->Pi);
         MemFree(ProblemeAResoudre->Colonne);
 
+        ProblemeAResoudre->NomDesVariables.clear();
```
Is it really useful (to deallocate vectors manually) ?

flomnes replied:
No, the underlying memory should be deallocated when `~std::vector<T>` is called. This code implies meaning but can be safely removed.

flomnes committed:
Rename PROBLEME_NOMME -> PROBLEME_SIMPLEXE_NOMME

flomnes committed:
Use const char* for NomDeLaContrainteCouplante

flomnes committed:
NomDesVariables/Contraintes: use const ref instead of copy

flomnes committed:
Rename ts -> tsInYear

flomnes committed:
Change type of BindingConstraintRTI::name to const char*

flomnes commented on src/solver/utils/ortools_utils.cpp:
```diff
@@ -12,31 +12,36 @@ using namespace operations_research;
+          x = solver->MakeNumVar(min_l, max_l, oss.str());
+        } else {
+          x = solver->MakeNumVar(min_l, max_l, NomDesVariables[idxVar]);
+        }
```
Use temporary for variable name => one single call to MakeNumVar

flomnes committed:
Replace SmartPtr<Study>& by SmartPtr<Study>

flomnes committed:
Remove failing asserts

They should fail since PROBLEME_SIMPLEXE has not yet been constructed
at instantiation.

flomnes committed:
Revert "Change type of BindingConstraintRTI::name to const char*"

This reverts commit e7c994e4228560d0dd0ad4180541dbe3f5201bae.

flomnes committed:
Use std::string for constraint name

flomnes committed:
Simplify impl of transferVariables & transferRows

flomnes committed:
Merge 'origin/develop' into feature/var_constr_naming

flomnes commented:
Contained in https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/431
